File: operators/network/transactions/bootstrap.mjs
/** External, bounded PoX-4 bootstrap. Never run against a reused environment. */
import { readFileSync, writeFileSync } from "node:fs";
import { execFileSync, spawn } from "node:child_process";
import { setTimeout as sleep } from "node:timers/promises";
import { StackingClient, Pox4SignatureTopic } from "@stacks/stacking";
import { localNetwork } from "./network.mjs";

const args = Object.fromEntries(
  process.argv.slice(2).map((value) => {
    const i = value.indexOf("=");
    if (i < 0) throw Error("use --option=value");
    return [value.slice(2, i), value.slice(i + 1)];
  }),
);
if (!args.manifest || !args.kubeconfig || !args.context)
  throw Error("--manifest, --kubeconfig and --context are required");
const doc = JSON.parse(readFileSync(args.manifest, "utf8")),
  setup = doc.bootstrap;
const prefix = [
  "--kubeconfig",
  args.kubeconfig,
  "--context",
  args.context,
  "-n",
  setup.namespace,
];
/** Execute a bounded command against the explicitly selected namespace. */
const kube = (...argv) =>
  execFileSync("kubectl", [...prefix, ...argv], {
    encoding: "utf8",
    timeout: 30000,
    maxBuffer: 2 << 20,
    stdio: ["pipe", "pipe", "pipe"],
  });
/** Read one network-scoped Kubernetes resource. */
const get = (resource) =>
  JSON.parse(kube("get", resource, setup.networkName, "-o", "json"));
/** Change only the external bootstrap's declared parent fields. */
const patch = (spec) =>
  kube(
    "patch",
    "stacksnetwork",
    setup.networkName,
    "--type=merge",
    "-p",
    JSON.stringify({ spec }),
  );
const forwarders = [];
const ports = {
  bitcoin: Number(args["bitcoin-port"] ?? 19443),
  stacks: Number(args["stacks-port"] ?? 20443),
};
const bitcoinURL = `http://127.0.0.1:${ports.bitcoin}`,
  stacksURL = `http://127.0.0.1:${ports.stacks}`;
const evidence = {
  schemaVersion: 1,
  namespace: setup.namespace,
  network: setup.networkName,
  events: [],
};
/** Persist credential-free bootstrap observations. */
const record = (event, details = {}) => {
  evidence.events.push({ event, at: new Date().toISOString(), ...details });
  writeFileSync(
    args.evidence ?? `${args.manifest}.bootstrap-evidence.json`,
    JSON.stringify(evidence, null, 2) + "\n",
    { mode: 0o600 },
  );
  console.log(event, JSON.stringify(details));
};
/** Poll a prerequisite within the given wall-clock bound. */
async function wait(label, fn, seconds = 180) {
  const deadline = Date.now() + seconds * 1000;
  while (Date.now() < deadline) {
    if (await fn()) return;
    await sleep(1000);
  }
  throw Error(`${label} did not complete`);
}
/** Open a loopback-only forwarding session owned by this helper. */
async function forward(service, local, remote) {
  const child = spawn(
    "kubectl",
    [
      ...prefix,
      "port-forward",
      "--address",
      "127.0.0.1",
      `service/${service}`,
      `${local}:${remote}`,
    ],
    { stdio: ["ignore", "pipe", "pipe"] },
  );
  forwarders.push(child);
  child.stderr.resume();
  await new Promise((resolve, reject) => {
    const timer = setTimeout(
      () => reject(Error("port-forward startup timed out")),
      20000,
    );
    child.stdout.on("data", (value) => {
      if (String(value).includes("Forwarding from")) {
        clearTimeout(timer);
        resolve();
      }
    });
    child.on("exit", () => {
      clearTimeout(timer);
      reject(Error("port-forward exited"));
    });
  });
}
/** Issue one Bitcoin request without retrying uncertain mutations. */
async function bitcoin(method, params = [], wallet = "") {
  const response = await fetch(
    bitcoinURL + (wallet ? `/wallet/${wallet}` : ""),
    {
      method: "POST",
      signal: AbortSignal.timeout(
        method === "generatetoaddress" ? 120000 : 30000,
      ),
      headers: {
        Authorization:
          "Basic " +
          Buffer.from(
            setup.bitcoin.username + ":" + setup.bitcoin.password,
          ).toString("base64"),
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ jsonrpc: "2.0", id: "bootstrap", method, params }),
    },
  );
  const value = await response.json();
  if (!response.ok || value.error)
    throw Error(`Bitcoin ${method} failed; do not blindly rerun bootstrap`);
  return value.result;
}
/** Read a bounded native Stacks response. */
async function json(path) {
  const response = await fetch(stacksURL + path, {
    signal: AbortSignal.timeout(5000),
  });
  if (!response.ok) throw Error("Stacks read unavailable");
  return response.json();
}
try {
  await wait("Bitcoin readiness", () => {
    try {
      return (
        JSON.parse(
          kube(
            "get",
            "bitcoinnode",
            setup.networkName + "-bitcoin",
            "-o",
            "json",
          ),
        ).status?.ready === true
      );
    } catch {
      return false;
    }
  });
  await forward(setup.networkName + "-bitcoin", ports.bitcoin, 18443);
  const parent = get("stacksnetwork"),
    ledger = get("bitcoinblockproduction");
  if (
    parent.spec.bitcoinBlockProduction?.paused !== true ||
    ledger.status?.dispatchID ||
    (await bitcoin("getblockcount")) !== 0
  )
    throw Error(
      "bootstrap requires a fresh height-zero network and a never-used paused Bitcoin ledger",
    );
  const wallet = await bitcoin("createwallet", [
    "stacks-miner",
    true,
    true,
    "",
    false,
    true,
  ]);
  if (!wallet.name) throw Error("wallet creation did not confirm");
  const descriptor = await bitcoin("getdescriptorinfo", [
    `addr(${setup.bitcoin.address})`,
  ]);
  const imported = await bitcoin(
    "importdescriptors",
    [[{ desc: descriptor.descriptor, timestamp: 0 }]],
    "stacks-miner",
  );
  if (!imported[0]?.success) throw Error("descriptor import did not confirm");
  const hashes = await bitcoin("generatetoaddress", [
    201,
    setup.bitcoin.address,
    1000000,
  ]);
  if (hashes.length !== 201) throw Error("bootstrap generation is incomplete");
  record("BitcoinBootstrap", {
    blocks: hashes.length,
    lastBlockHash: hashes.at(-1),
  });
  patch({
    stacksNodes: parent.spec.stacksNodes.map((x) => ({
      ...x,
      suspended: false,
    })),
    signers: parent.spec.signers.map((x) => ({ ...x, suspended: false })),
  });
  await wait(
    "Stacks ingress readiness",
    () => {
      try {
        return (
          JSON.parse(
            kube(
              "get",
              "stacksnode",
              setup.networkName + "-signer-node",
              "-o",
              "json",
            ),
          ).status?.ready === true
        );
      } catch {
        return false;
      }
    },
    240,
  );
  await forward(setup.networkName + "-signer-node", ports.stacks, 20443);
  await wait(
    "Stacks burnchain catchup",
    async () => {
      try {
        return (await json("/v2/info")).burn_block_height >= 201;
      } catch {
        return false;
      }
    },
    180,
  );
  patch({ bitcoinBlockProduction: { paused: false } });
  const network = localNetwork(stacksURL);
  const client = new StackingClient({ address: setup.signer.address, network });
  await wait(
    "PoX-4 activation",
    async () => {
      try {
        return (await client.getPoxInfo()).contract_id.endsWith(".pox-4");
      } catch {
        return false;
      }
    },
    180,
  );
  const info = await client.getPoxInfo(),
    status = await client.getAccountStatus();
  const amount = BigInt(Math.floor(info.next_cycle.min_threshold_ustx * 1.5)),
    authId = 1n,
    maxAmount = (1n << 128n) - 1n,
    cycles = 12;
  if (BigInt(status.balance) < amount)
    throw Error("signer genesis balance is insufficient");
  const signature = client.signPoxSignature({
    topic: Pox4SignatureTopic.StackStx,
    rewardCycle: info.reward_cycle_id,
    poxAddress: setup.signer.poxAddress,
    period: cycles,
    signerPrivateKey: setup.signer.privateKey,
    authId,
    maxAmount,
  });
  const result = await client.stack({
    poxAddress: setup.signer.poxAddress,
    privateKey: setup.signer.privateKey,
    amountMicroStx: amount,
    burnBlockHeight: info.current_burnchain_block_height,
    cycles,
    fee: 3000,
    nonce: status.nonce,
    signerKey: setup.signer.publicKey,
    signerSignature: signature,
    authId,
    maxAmount,
  });
  if (!result.txid)
    throw Error(
      "enrollment submission did not acknowledge a transaction; inspect its nonce before any further action",
    );
  record("SignerEnrollmentSubmitted", { txid: result.txid });
  await wait(
    "signer lock confirmation",
    async () => BigInt((await client.getAccountStatus()).locked) > 0n,
    180,
  );
  record("SignerEnrollmentLocked", {
    unlockHeight: (await client.getAccountStatus()).unlock_height,
  });
  await wait(
    "Nakamoto activation",
    async () => {
      const value = await json("/v2/info");
      return value.burn_block_height >= 230 && value.stacks_tip_height > 0;
    },
    240,
  );
  patch({ stacksTransactionProduction: { paused: false } });
  await wait(
    "first exact transfer inclusion",
    () => get("stackstransactionproduction").status?.confirmed > 0,
    240,
  );
  const ready = await json("/v2/info");
  record("BootstrapComplete", {
    burnHeight: ready.burn_block_height,
    stacksHeight: ready.stacks_tip_height,
    serverVersion: ready.server_version,
    genesisHash: ready.genesis_chainstate_hash,
    confirmedTransfers: get("stackstransactionproduction").status.confirmed,
  });
} finally {
  for (const child of forwarders) child.kill("SIGTERM");
}

File: operators/network/internal/bootstrap/bootstrap.go
package bootstrap

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"reflect"
	"strconv"

	bitcoin "github.com/cylewitruk-stacks/stacks-k8s/apis/network/bitcoin/v1alpha1"
	stacks "github.com/cylewitruk-stacks/stacks-k8s/apis/network/stacks/v1alpha1"
	network "github.com/cylewitruk-stacks/stacks-k8s/apis/network/v1alpha1"
	"github.com/cylewitruk-stacks/stacks-k8s/operators/network/internal/environment"
	"github.com/cylewitruk-stacks/stacks-k8s/operators/network/internal/naming"
	"github.com/cylewitruk-stacks/stacks-k8s/operators/network/internal/profiles"
)

// manifest reads only the network and private bootstrap binding from the provisioning document.
type manifest struct {
	Items     []json.RawMessage     `json:"items"`
	Bootstrap environment.Bootstrap `json:"bootstrap"`
}

// Load validates the private provisioning document and its immutable genesis binding.
func Load(path string) (environment.Bootstrap, *network.StacksNetwork, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return environment.Bootstrap{}, nil, err
	}
	var doc manifest
	if json.Unmarshal(data, &doc) != nil {
		return doc.Bootstrap, nil, fmt.Errorf("invalid provisioning document")
	}
	var parent *network.StacksNetwork
	for _, raw := range doc.Items {
		var header struct {
			Kind string `json:"kind"`
		}
		if json.Unmarshal(raw, &header) != nil {
			return doc.Bootstrap, nil, fmt.Errorf("invalid resource")
		}
		if header.Kind == "StacksNetwork" {
			if parent != nil {
				return doc.Bootstrap, nil, fmt.Errorf("expected one network")
			}
			parent = &network.StacksNetwork{}
			if err = json.Unmarshal(raw, parent); err != nil {
				return doc.Bootstrap, nil, fmt.Errorf("invalid network")
			}
		}
	}
	if parent == nil || parent.Name != doc.Bootstrap.NetworkName || parent.Namespace != doc.Bootstrap.Namespace || parent.Spec.Genesis == nil {
		return doc.Bootstrap, nil, fmt.Errorf("bootstrap/network identity mismatch")
	}
	if err = ValidateGenesis(doc.Bootstrap, parent); err != nil {
		return doc.Bootstrap, nil, err
	}
	return doc.Bootstrap, parent, nil
}

// ValidateGenesis verifies the shared snapshot and named bootstrap account before any mutation.
func ValidateGenesis(setup environment.Bootstrap, parent *network.StacksNetwork) error {
	digest, err := environment.GenesisDigest(parent.Spec.Genesis)
	if err != nil {
		return err
	}
	if digest != setup.GenesisDigest {
		return fmt.Errorf("bootstrap genesis digest mismatch")
	}
	found := false
	for _, a := range parent.Spec.Genesis.Balances {
		if a.Name == setup.SignerAccount && a.Address == setup.Signer.Address {
			found = true
		}
	}
	if !found {
		return fmt.Errorf("bootstrap account is absent from network genesis")
	}
	expected := profiles.DefaultGenesis()
	resolved, err := profiles.ResolveGenesis(parent.Spec.Genesis)
	if err != nil {
		return err
	}
	if !reflect.DeepEqual(resolved.UseTestGenesisChainstate, expected.UseTestGenesisChainstate) || !reflect.DeepEqual(resolved.Epochs, expected.Epochs) || !reflect.DeepEqual(resolved.PoX, expected.PoX) {
		return fmt.Errorf("external bootstrap currently supports only the qualified PoX-4 regtest schedule")
	}
	return nil
}

// Run initializes a fresh suspended network, enrolls its seeded stacker and verifies first production.
func Run(ctx context.Context, options Options) error {
	setup, declared, err := Load(options.Manifest)
	if err != nil {
		return err
	}
	if options.Kubeconfig == "" || options.Context == "" || options.BitcoinPort < 1 || options.BitcoinPort > 65535 || options.StacksPort < 1 || options.StacksPort > 65535 || options.StacksPort == options.BitcoinPort {
		return fmt.Errorf("explicit kubeconfig/context and distinct valid loopback ports are required")
	}
	s := &session{options: options, namespace: setup.Namespace, http: &http.Client{Transport: &http.Transport{DisableKeepAlives: true}, CheckRedirect: func(*http.Request, []*http.Request) error { return http.ErrUseLastResponse }}}
	if err = s.claimEvidence("bootstrap"); err != nil {
		return err
	}
	defer s.close()
	ready := func(kind, name string) bool {
		var actor struct {
			Status struct {
				Ready bool `json:"ready"`
			} `json:"status"`
		}
		return s.get(ctx, kind, name, &actor) == nil && actor.Status.Ready
	}
	if err = wait(ctx, "Bitcoin readiness", 180, func() bool { return ready("bitcoinnode", naming.Child(setup.NetworkName, "bitcoin")) }); err != nil {
		return err
	}
	if err = s.forward(ctx, naming.Child(setup.NetworkName, "bitcoin"), options.BitcoinPort, 18443); err != nil {
		return err
	}
	var parent network.StacksNetwork
	if err = s.get(ctx, "stacksnetwork", setup.NetworkName, &parent); err != nil {
		return err
	}
	if err = ValidateGenesis(setup, &parent); err != nil {
		return err
	}
	if !reflect.DeepEqual(parent.Spec.BitcoinBlockProduction, declared.Spec.BitcoinBlockProduction) || !reflect.DeepEqual(parent.Spec.StacksNodes, declared.Spec.StacksNodes) || !reflect.DeepEqual(parent.Spec.Signers, declared.Spec.Signers) || !reflect.DeepEqual(parent.Spec.BitcoinNodes, declared.Spec.BitcoinNodes) || !reflect.DeepEqual(parent.Spec.StacksTransactionProduction, declared.Spec.StacksTransactionProduction) {
		return fmt.Errorf("bootstrap actor declarations changed")
	}
	var ledger bitcoin.BitcoinProductionTarget
	if err = s.get(ctx, "bitcoinproductiontarget", naming.Child(setup.NetworkName, "bitcoin"), &ledger); err != nil {
		return err
	}
	var height int64
	if err = s.bitcoin(ctx, setup, "getblockcount", nil, "", &height); err != nil {
		return err
	}
	if err = validateFreshBitcoin(&parent, &ledger, height); err != nil {
		return err
	}

	if parent.Spec.StacksTransactionProduction == nil || !parent.Spec.StacksTransactionProduction.Paused {
		return fmt.Errorf("bootstrap requires paused transfer production")
	}
	for _, node := range parent.Spec.StacksNodes {
		if !node.Suspended {
			return fmt.Errorf("bootstrap requires initially suspended Stacks nodes")
		}
	}
	for _, signer := range parent.Spec.Signers {
		if !signer.Suspended {
			return fmt.Errorf("bootstrap requires initially suspended signers")
		}
	}
	// Persist the experiment identity before mutation; failure is not automatically replayed.
	if err = s.record("BootstrapStarted", map[string]any{"network": parent.Name, "networkUID": string(parent.UID), "genesisDigest": setup.GenesisDigest}); err != nil {
		return err
	}
	var wallet struct {
		Name string `json:"name"`
	}
	if err = s.bitcoin(ctx, setup, "createwallet", []any{"stacks-miner", true, true, "", false, true}, "", &wallet); err != nil {
		return err
	}
	if wallet.Name == "" {
		return fmt.Errorf("wallet creation did not confirm")
	}
	var descriptor struct {
		Descriptor string `json:"descriptor"`
	}
	if err = s.bitcoin(ctx, setup, "getdescriptorinfo", []any{"addr(" + setup.Bitcoin.Address + ")"}, "", &descriptor); err != nil {
		return err
	}
	var imported []struct {
		Success bool `json:"success"`
	}
	if err = s.bitcoin(ctx, setup, "importdescriptors", []any{[]any{map[string]any{"desc": descriptor.Descriptor, "timestamp": 0}}}, "stacks-miner", &imported); err != nil {
		return err
	}
	if len(imported) != 1 || !imported[0].Success {
		return fmt.Errorf("descriptor import did not confirm")
	}
	var hashes []string
	if err = s.bitcoin(ctx, setup, "generatetoaddress", []any{201, setup.Bitcoin.Address, 1000000}, "", &hashes); err != nil {
		return err
	}
	if len(hashes) != 201 {
		return fmt.Errorf("bootstrap generation is incomplete")
	}
	if err = s.record("BitcoinBootstrap", map[string]any{"blocks": len(hashes), "lastBlockHash": hashes[len(hashes)-1]}); err != nil {
		return err
	}
	for i := range parent.Spec.StacksNodes {
		parent.Spec.StacksNodes[i].Suspended = false
	}
	for i := range parent.Spec.Signers {
		parent.Spec.Signers[i].Suspended = false
	}
	if err = s.patch(ctx, parent.Name, map[string]any{"stacksNodes": parent.Spec.StacksNodes, "signers": parent.Spec.Signers}); err != nil {
		return err
	}
	if err = wait(ctx, "Stacks ingress readiness", 240, func() bool { return ready("stacksnode", naming.Child(parent.Name, "signer-node")) }); err != nil {
		return err
	}
	if err = s.forward(ctx, naming.Child(parent.Name, "signer-node"), options.StacksPort, 20443); err != nil {
		return err
	}
	var info nodeInfo
	if err = wait(ctx, "Stacks burnchain catchup", 180, func() bool { return s.stacks(ctx, "/v2/info", &info) == nil && info.BurnHeight >= 201 }); err != nil {
		return err
	}
	if err = s.patch(ctx, parent.Name, map[string]any{"bitcoinBlockProduction": map[string]any{"paused": false}}); err != nil {
		return err
	}
	var pox poxInfo
	if err = wait(ctx, "PoX-4 activation", 180, func() bool { return s.stacks(ctx, "/v2/pox", &pox) == nil && pox.Contract == pox4 }); err != nil {
		return err
	}
	account, err := s.account(ctx, setup.Signer.Address)
	if err != nil {
		return err
	}
	amount := pox.NextCycle.MinThreshold + pox.NextCycle.MinThreshold/2
	if amount < 1 || account.Balance < amount {
		return fmt.Errorf("signer genesis balance is insufficient")
	}
	tx, err := s.sign(ctx, setup, "enroll", pox, account, amount, 12, 1)
	if err != nil {
		return err
	}
	if err = s.record("SignerEnrollmentAuthorized", map[string]any{"txid": tx.TxID, "nonce": account.Nonce}); err != nil {
		return err
	}
	if err = s.submit(ctx, tx); err != nil {
		return err
	}
	if err = s.record("SignerEnrollmentSubmitted", map[string]any{"txid": tx.TxID}); err != nil {
		return err
	}
	if err = wait(ctx, "signer lock confirmation", 180, func() bool {
		var readErr error
		account, readErr = s.account(ctx, setup.Signer.Address)
		return readErr == nil && account.Locked > 0
	}); err != nil {
		return err
	}
	if err = s.record("SignerEnrollmentLocked", map[string]any{"unlockHeight": account.UnlockHeight}); err != nil {
		return err
	}
	if err = wait(ctx, "Nakamoto activation", 240, func() bool {
		return s.stacks(ctx, "/v2/info", &info) == nil && info.BurnHeight >= 230 && info.StacksHeight > 0
	}); err != nil {
		return err
	}
	if err = s.patch(ctx, parent.Name, map[string]any{"stacksTransactionProduction": map[string]any{"paused": false}}); err != nil {
		return err
	}
	var production stacks.StacksTransactionProduction
	if err = wait(ctx, "first exact transfer inclusion", 240, func() bool {
		return s.get(ctx, "stackstransactionproduction", parent.Name, &production) == nil && production.Status.Confirmed > 0
	}); err != nil {
		return err
	}
	if err = s.stacks(ctx, "/v2/info", &info); err != nil {
		return err
	}
	return s.record("BootstrapComplete", map[string]any{"burnHeight": info.BurnHeight, "stacksHeight": info.StacksHeight, "serverVersion": info.ServerVersion, "genesisHash": info.GenesisHash, "confirmedTransfers": production.Status.Confirmed})
}

// bitcoin performs one RPC attempt and decodes only acknowledged success.
func (s *session) bitcoin(ctx context.Context, setup environment.Bootstrap, method string, params any, wallet string, out any) error {
	if params == nil {
		params = []any{}
	}
	body, _ := json.Marshal(map[string]any{"jsonrpc": "2.0", "id": "bootstrap", "method": method, "params": params})
	url := "http://127.0.0.1:" + strconv.Itoa(s.options.BitcoinPort)
	if wallet != "" {
		url += "/wallet/" + wallet
	}
	seconds := 30
	if method == "generatetoaddress" {
		seconds = 120
	}
	data, err := s.request(ctx, "POST", url, body, setup.Bitcoin.Username, setup.Bitcoin.Password, "application/json", seconds)
	if err != nil {
		return err
	}
	var result struct {
		Result json.RawMessage `json:"result"`
		Error  json.RawMessage `json:"error"`
	}
	if json.Unmarshal(data, &result) != nil || (len(result.Error) > 0 && string(result.Error) != "null") || len(result.Result) == 0 || string(result.Result) == "null" {
		return fmt.Errorf("Bitcoin %s did not acknowledge success", method)
	}
	if json.Unmarshal(result.Result, out) != nil {
		return fmt.Errorf("invalid Bitcoin %s receipt", method)
	}
	return nil
}

// nodeInfo retains only readiness and identity evidence needed by bootstrap.
type nodeInfo struct {
	BurnHeight    int64  `json:"burn_block_height"`
	StacksHeight  int64  `json:"stacks_tip_height"`
	ServerVersion string `json:"server_version"`
	GenesisHash   string `json:"genesis_chainstate_hash"`
}

// stacks reads one bounded native Stacks endpoint.
func (s *session) stacks(ctx context.Context, path string, out any) error {
	data, err := s.request(ctx, "GET", "http://127.0.0.1:"+strconv.Itoa(s.options.StacksPort)+path, nil, "", "", "", 5)
	if err != nil {
		return err
	}
	if json.Unmarshal(data, out) != nil {
		return fmt.Errorf("invalid Stacks read response")
	}
	return nil
}

// validateFreshBitcoin excludes reused, foreign, unpaused or action-reserved ledgers.
func validateFreshBitcoin(parent *network.StacksNetwork, ledger *bitcoin.BitcoinProductionTarget, height int64) error {
	if parent.UID == "" || !parent.DeletionTimestamp.IsZero() || !ledger.DeletionTimestamp.IsZero() || parent.Spec.BitcoinBlockProduction == nil || !parent.Spec.BitcoinBlockProduction.Paused || ledger.Spec.NetworkUID != string(parent.UID) || ledger.Spec.NetworkName != parent.Name || ledger.Spec.ProductionUID != parent.Status.BitcoinProductionUID || ledger.Spec.Policy.Target != "bitcoin" || !ledger.Spec.Policy.Paused || ledger.Status.DispatchID != "" || ledger.Status.DispatchState != "" || ledger.Status.BlocksProduced != 0 || ledger.Status.Action != nil || ledger.Status.Reorganization != nil || height != 0 {
		return fmt.Errorf("bootstrap requires a fresh height-zero network and its unused, unreserved paused Bitcoin ledger")
	}
	return nil
}
